From 89bbc3d241f7f1de6dc497b58bd347613b495514 Mon Sep 17 00:00:00 2001
From: Stefan Berger <stefanb@linux.ibm.com>
Date: Wed, 16 Mar 2022 12:32:41 -0400
Subject: [PATCH 36/81] ima: Restrict informational audit messages to
 init_ima_ns

Restrict informational audit messages that may be emitted due to IMA
measurement being active in an IMA namespace to the init_ima_ns.

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>
---
 security/integrity/ima/ima_api.c  | 13 ++++++++-----
 security/integrity/ima/ima_main.c |  2 +-
 2 files changed, 9 insertions(+), 6 deletions(-)

diff --git a/security/integrity/ima/ima_api.c b/security/integrity/ima/ima_api.c
index be96aa18c8d5..c2b9e01c4beb 100644
--- a/security/integrity/ima/ima_api.c
+++ b/security/integrity/ima/ima_api.c
@@ -118,9 +118,10 @@ int ima_store_template(struct ima_namespace *ns,
 						   &entry->template_data[0],
 						   entry);
 		if (result < 0) {
-			integrity_audit_msg(AUDIT_INTEGRITY_PCR, inode,
-					    template_name, op,
-					    audit_cause, result, 0);
+			if (ns == &init_ima_ns)
+				integrity_audit_msg(AUDIT_INTEGRITY_PCR, inode,
+						    template_name, op,
+						    audit_cause, result, 0);
 			return result;
 		}
 	}
@@ -428,8 +429,10 @@ void ima_store_measurement(struct ima_namespace *ns,
 
 	result = ima_alloc_init_template(ns, &event_data, &entry, template_desc);
 	if (result < 0) {
-		integrity_audit_msg(AUDIT_INTEGRITY_PCR, inode, filename,
-				    op, audit_cause, result, 0);
+		if (ns == &init_ima_ns)
+			integrity_audit_msg(AUDIT_INTEGRITY_PCR, inode,
+					    filename, op, audit_cause, result,
+					    0);
 		return;
 	}
 
diff --git a/security/integrity/ima/ima_main.c b/security/integrity/ima/ima_main.c
index e69b0742bd4d..2aa12ce160c6 100644
--- a/security/integrity/ima/ima_main.c
+++ b/security/integrity/ima/ima_main.c
@@ -1106,7 +1106,7 @@ int process_buffer_measurement(struct ima_namespace *ns,
 	}
 
 out:
-	if (ret < 0)
+	if (ret < 0 && ns == &init_ima_ns)
 		integrity_audit_message(AUDIT_INTEGRITY_PCR, NULL, eventname,
 					func_measure_str(func),
 					audit_cause, ret, 0, ret);
-- 
2.36.1

